compiler/aarch64/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5624,10 +5624,358 @@ OMR::ARM64::TreeEvaluator::arraysetEvaluator(TR::Node *node, TR::CodeGenerator *
+    * cmp     data1Reg, #0
+    * csel    data1Reg, data1Reg, data3Reg, ne
+    * mov     data2Reg, #1
+    * cinc    data2Reg, data2Reg, ne
+    * sub     src1Reg, src1Reg, data2Reg, lsl #3 ; Adjusts post incremented address
+    * rbit    data1Reg, data1Reg
+    * clz     data1Reg, data1Reg
+    * add     src1Reg, src1Reg, data1Reg, lsr #3
+    *
+    * ;     non-arrayLen case
+    * cmp     data1Reg, data2Reg
+    * csel    data1Reg, data1Reg, data3Reg, ne
+    * csel    data2Reg, data2Reg, data4Reg, ne
+    * rev64   data1Reg, data1Reg
+    * rev64   data2Reg, data2Reg
+    *
+    * if !(length is constant and length > 15) {
+    *    b       LDONE0
+    *
+    *    LessThan16:
+    *    byteloop:
+    *    subs    lengthReg, lengthReg, #1
+    *    ldrb    data1Reg, [src1Reg], #1
+    *    ldrb    data2Reg, [src2Reg], #1
+    *    ccmp    data1Reg, data2Reg, #0, hi
+    *    b.eq    byteloop
+    *    ;     arrayLen case
+    *    cmp     data1Reg, data2Reg
+    *    cset    offReg, ne
+    *    sub     src1Reg, src1Reg, offReg
+    * }
+    * ;     arrayLen case
+    * LDONE0:
+    * sub     resultReg, src1Reg, savedSrc1Reg
+    * ;     non-arrayLen case
+    * LDONE0:
+    * cmp     data1Reg, data2Reg
+    * cset    resultReg, ne
+    * cinc    resultReg, resultReg, hi     ; Returns 1 or 2
+    *
+    * LDONE:
+    */
+   TR::Node *src1Node = node->getFirstChild();
+   TR::Node *src2Node = node->getSecondChild();
+   TR::Node *lengthNode = node->getThirdChild();
+   bool isLengthGreaterThan15 = lengthNode->getOpCode().isLoadConst() && lengthNode->getConstValue() > 15;
+   const bool isArrayCmpLen = node->isArrayCmpLen();
+   TR_ARM64ScratchRegisterManager *srm = cg->generateScratchRegisterManager(12);
+
+   TR::Register *savedSrc1Reg = cg->evaluate(src1Node);
+   TR::Register *src1Reg;
+   if ((src1Node->getReferenceCount() > 1) || isArrayCmpLen)
+      {
+      src1Reg = srm->findOrCreateScratchRegister();
+      generateMovInstruction(cg, node, src1Reg, savedSrc1Reg);
+      }
+   else
+      {
+      src1Reg = savedSrc1Reg;
+      }
+   TR::Register *src2Reg = cg->gprClobberEvaluate(src2Node);
+   TR::Register *lengthReg = cg->gprClobberEvaluate(lengthNode);
+   TR::Register *resultReg = cg->allocateRegister();
+   TR::LabelSymbol *startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *doneLabel = generateLabelSymbol(cg);
+   TR_Debug * debugObj = cg->getDebug();
+
+   startLabel->setStartInternalControlFlow();
+   doneLabel->setEndInternalControlFlow();
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+   if (isArrayCmpLen)
+      {
+      generateMovInstruction(cg, node, resultReg, lengthReg, false);
+      }
+   else
+      {
+      loadConstant32(cg, node, 0, resultReg);
+      }
+   generateCompareInstruction(cg, node, src1Reg, src2Reg, true);
+   if (!isLengthGreaterThan15)
+      {
+      auto ccmpLengthInstr = generateConditionalCompareImmInstruction(cg, node, lengthReg, 0, 4, TR::CC_NE);
```

Comment:
Please add a comment like `/* 4 for Z flag */'.

Reply:
Added the comment.